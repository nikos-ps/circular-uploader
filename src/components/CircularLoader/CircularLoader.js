import React, { useMemo } from "react";

import { Container, Label, RotatingCircle } from "./CircularLoader.styled";

const CircularLoader = ({ radius, stroke, progress }) => {
  const normalizedRadius = useMemo(() => radius - stroke / 2, [radius, stroke]);
  const circumference = useMemo(
    () => normalizedRadius * 2 * Math.PI,
    [normalizedRadius]
  );
  const strokeDashoffset = circumference - (progress / 100) * circumference;

  return (
    <Container>
      <svg height={radius * 2} width={radius * 2} data-testid="circular-loader">
        <circle
          stroke="var(--colors-porcelain)"
          fill="transparent"
          strokeWidth={stroke}
          strokeDasharray={circumference + " " + circumference}
          r={normalizedRadius}
          cx={radius}
          cy={radius}
        ></circle>
        <RotatingCircle
          stroke="var(--colors-dodgerBlue)"
          fill="transparent"
          strokeWidth={stroke}
          strokeDasharray={circumference + " " + circumference}
          strokeDashoffset={strokeDashoffset}
          r={normalizedRadius}
          cx={radius}
          cy={radius}
        />
      </svg>
      <Label data-testid="progress-label">
        {progress}
        <span>%</span>
      </Label>
    </Container>
  );
};

export default CircularLoader;
