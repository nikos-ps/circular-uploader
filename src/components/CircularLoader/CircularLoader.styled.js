import styled, { keyframes } from "styled-components";

export const Container = styled.div`
  display: block;
  margin: 30px auto 12px;
  position: relative;
  width: 170px;
`;

const rotation = keyframes`
   from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(1turn);
  }
`;

export const RotatingCircle = styled.circle`
  border-radius: 10px;
  stroke-linecap: round;
  transition: stroke-dashoffset 0.2s linear, stroke 0.2s linear;
  transform: rotate(-90deg);
  transform-origin: 50% 50%;
  animation: ${rotation} 2s linear infinite;
`;

export const Label = styled.span`
  color: var(--colors-woodSmoke);
  font-size: 3.875em;
  left: 50%;
  line-height: 1;
  margin: -2px 0 0 2px;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);
  white-space: nowrap;

  span {
    color: var(--colors-silverSand);
    display: inline-block;
    font-size: 22px;
    line-height: 1;
    position: relative;
    top: 9px;
    vertical-align: top;
  }
`;
