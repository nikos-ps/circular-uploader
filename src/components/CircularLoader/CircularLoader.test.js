import { render } from "@testing-library/react";

import CircularLoader from "./CircularLoader";

describe("<CircularLoader />", () => {
  const radius = 85;
  const stroke = 10;

  const renderLoader = (progress = 0) => {
    return render(
      <CircularLoader radius={radius} stroke={stroke} progress={progress} />
    );
  };

  it("renders correctly", () => {
    const { container } = renderLoader();

    expect(container).toMatchSnapshot();
  });

  it("renders an svg element correctly", () => {
    const { container } = renderLoader();

    const svgElement = container.querySelector("svg");
    expect(svgElement).toBeInTheDocument();

    const expected = (2 * radius).toString();
    expect(svgElement).toHaveAttribute("width", expected);
    expect(svgElement).toHaveAttribute("height", expected);
  });

  it("renders the background circle correctly", () => {
    const { container } = renderLoader();

    const bgCircle = container.querySelectorAll("circle")[0];
    const { normalizedRadius, circumference } = getCircleAttributes(
      radius,
      stroke
    );

    expect(bgCircle).toHaveAttribute("stroke-width", String(stroke));
    expect(bgCircle).toHaveAttribute("cx", String(radius));
    expect(bgCircle).toHaveAttribute("cy", String(radius));
    expect(bgCircle).toHaveAttribute("r", String(normalizedRadius));
    expect(bgCircle).toHaveAttribute(
      "stroke-dasharray",
      `${circumference} ${circumference}`
    );
  });

  it("renders the rotating circle correctly", () => {
    const { container } = renderLoader();

    const rotatingCircle = container.querySelectorAll("circle")[1];
    const { normalizedRadius, circumference, strokeDashoffset } =
      getCircleAttributes(radius, stroke);

    expect(rotatingCircle).toHaveAttribute("stroke-width", String(stroke));
    expect(rotatingCircle).toHaveAttribute("cx", String(radius));
    expect(rotatingCircle).toHaveAttribute("cy", String(radius));
    expect(rotatingCircle).toHaveAttribute("r", String(normalizedRadius));
    expect(rotatingCircle).toHaveAttribute(
      "stroke-dasharray",
      `${circumference} ${circumference}`
    );
    expect(rotatingCircle).toHaveAttribute(
      "stroke-dashoffset",
      String(strokeDashoffset)
    );
  });

  test("the rotating circle's 'stroke-dashoffset' attribute is updating when progress has changed", () => {
    const { container, rerender } = renderLoader();

    const rotatingCircle = container.querySelectorAll("circle")[1];
    const { circumference, strokeDashoffset } = getCircleAttributes(
      radius,
      stroke
    );

    expect(rotatingCircle).toHaveAttribute(
      "stroke-dashoffset",
      String(strokeDashoffset)
    );

    const progress = 40;
    rerender(
      <CircularLoader radius={radius} stroke={stroke} progress={progress} />
    );

    const strokeDashoffsetUpdated =
      circumference - (progress / 100) * circumference;

    expect(rotatingCircle).toHaveAttribute(
      "stroke-dashoffset",
      String(strokeDashoffsetUpdated)
    );
  });

  it("renders the progress label correctly", () => {
    const progress = 40;
    const { getByTestId } = renderLoader(40);

    expect(getByTestId("progress-label")).toHaveTextContent(progress);
  });
});

const getCircleAttributes = (radius, stroke) => {
  const normalizedRadius = radius - stroke / 2;
  const circumference = normalizedRadius * 2 * Math.PI;
  const strokeDashoffset = circumference - (0 / 100) * circumference;

  return {
    normalizedRadius,
    circumference,
    strokeDashoffset,
  };
};
