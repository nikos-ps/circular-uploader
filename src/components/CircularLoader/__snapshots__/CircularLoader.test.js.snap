// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CircularLoader /> renders correctly 1`] = `
.c0 {
  display: block;
  margin: 30px auto 12px;
  position: relative;
  width: 170px;
}

.c1 {
  border-radius: 10px;
  stroke-linecap: round;
  -webkit-transition: stroke-dashoffset 0.2s linear,stroke 0.2s linear;
  transition: stroke-dashoffset 0.2s linear,stroke 0.2s linear;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  -webkit-animation: iwIftm 2s linear infinite;
  animation: iwIftm 2s linear infinite;
}

.c2 {
  color: var(--colors-woodSmoke);
  font-size: 3.875em;
  left: 50%;
  line-height: 1;
  margin: -2px 0 0 2px;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  white-space: nowrap;
}

.c2 span {
  color: var(--colors-silverSand);
  display: inline-block;
  font-size: 22px;
  line-height: 1;
  position: relative;
  top: 9px;
  vertical-align: top;
}

<div>
  <div
    class="c0"
  >
    <svg
      data-testid="circular-loader"
      height="170"
      width="170"
    >
      <circle
        cx="85"
        cy="85"
        fill="transparent"
        r="80"
        stroke="var(--colors-porcelain)"
        stroke-dasharray="502.6548245743669 502.6548245743669"
        stroke-width="10"
      />
      <circle
        class="c1"
        cx="85"
        cy="85"
        fill="transparent"
        r="80"
        stroke="var(--colors-dodgerBlue)"
        stroke-dasharray="502.6548245743669 502.6548245743669"
        stroke-dashoffset="502.6548245743669"
        stroke-width="10"
      />
    </svg>
    <span
      class="c2"
      data-testid="progress-label"
    >
      0
      <span>
        %
      </span>
    </span>
  </div>
</div>
`;
